CSharpMath/Structures/Dictionary.cs

```diff
@@ -0,0 +1,187 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+
+namespace CSharpMath.Structures {
+  /// <summary>
+  /// Funnels collection initializers to <see cref="Added"/>.
+  /// Implements <see cref="IEnumerable"/> but throws upon enumeration because it is there only to enable collection initializers.
+  /// </summary>
+  /// <example>
+  /// <code>new ProxyAdder&lt;int, string&gt;((key, value) => Console.WriteLine(value))
+  /// { { 1, 2, 3, "1 to 3" }, { Enumerable.Range(7, 10), i => i.ToString() } }</code>
+  /// </example>
+  [SuppressMessage("Naming", "CA1710:Identifiers should have correct suffix", Justification = NotACollection)]
+  [SuppressMessage("Design", "CA1010:Collections should implement generic interface", Justification = NotACollection)]
+  public class ProxyAdder<TKey, TValue> : IEnumerable {
+    const string NotACollection = "This is not a collection. It implements IEnumerable just to support collection initializers.";
+    [Obsolete(NotACollection, true)]
+    [SuppressMessage("Design", "CA1033:Interface methods should be callable by child types", Justification = NotACollection)]
+    IEnumerator IEnumerable.GetEnumerator() => throw new NotSupportedException(NotACollection);
+    public ProxyAdder(Action<TKey, TValue>? added = null) => Added += added;
+    public event Action<TKey, TValue>? Added;
+    public void Add(TKey key1, TValue value) => Added?.Invoke(key1, value);
+    public void Add(TKey key1, TKey key2, TValue value) {
+      Add(key1, value); Add(key2, value);
+    }
+    public void Add(TKey key1, TKey key2, TKey key3, TValue value) {
+      Add(key1, value); Add(key2, value); Add(key3, value);
+    }
+    public void Add<TCollection>(TCollection keys, TValue value) where TCollection : IEnumerable<TKey> {
+      foreach (var key in keys) Add(key, value);
+    }
+    public void Add<TCollection>(TCollection keys, Func<TKey, TValue> valueFunc) where TCollection : IEnumerable<TKey> {
+      foreach (var key in keys) Add(key, valueFunc(key));
+    }
+  }
+  [SuppressMessage("Naming", "CA1710:Identifiers should have correct suffix",
+    Justification = "This is conceptually a dictionary but has different lookup behavior")]
+  public class LaTeXCommandDictionary<TValue> : ProxyAdder<string, TValue>, IEnumerable<KeyValuePair<ReadOnlyMemory<char>, TValue>> {
```

Comment:
This type needs basic documentation. I.e. what is it, what is TValue.

Reply:
Copied from https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=netcore-3.1#type-parameters:

The type of the values in the dictionary.

Reply:
@Happypig375 this is nowhere near an explanation of what this type is.

Reply:
A dictionary that maps LaTeX commands and character symbols to values.